print-ospf.c

```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
+	    return -1;
+	}
+
+	ND_PRINT("\n\t      %s subTLV (%u), length: %u, value: ",
+		 tok2str(lsa_opaque_ri_sid_subtlv_values,"unknown",subtlv_type),
+		 subtlv_type,
+		 subtlv_length);
+
+	switch (subtlv_type) {
+	case LS_OPAQUE_RI_SUBTLV_SID_LABEL:
+	    if (subtlv_length == 3) {
+		ND_PRINT("\n\t\tLabel: %u", GET_BE_U_3(tptr));
+	    } else if (subtlv_length == 4) {
+		ND_PRINT("\n\t\tSID: %u", GET_BE_U_4(tptr));
+	    } else {
+		ND_PRINT("\n\t\tBogus subTLV length %u", subtlv_length);
+	    }
+	    break;
+
+	default:
+	    if (ndo->ndo_vflag <= 1) {
+		if (!print_unknown_data(ndo, tptr, "\n\t\t", subtlv_length))
+		    return -1;
+	    }
+	}
+
+	/* in OSPF everything has to be 32-bit aligned, including subTLVs */
+	if (subtlv_length % 4) {
+	    subtlv_length += (4 - (subtlv_length % 4));
+	}
+	tptr+=subtlv_length;
+	tlv_length-=subtlv_length;
+    }
+    return 0;
+trunc:
+    return -1;
+}
+
+static int
+ospf_print_ep_lsa_extd_prefix_tlv(netdissect_options *ndo, const uint8_t *tptr, u_int tlv_length)
+{
+    u_int subtlv_type, subtlv_length;
+    uint8_t flags, mt_id, algo;
+
+    while (tlv_length >= 4) {
+	ND_TCHECK_4(tptr);
+	subtlv_type = GET_BE_U_2(tptr);
+	subtlv_length = GET_BE_U_2(tptr+2);
+	tptr+=4;
+	tlv_length-=4;
+
+	/* Infinite loop protection. */
+	if (subtlv_type == 0 || subtlv_length == 0) {
+	    return -1;
+	}
+
+	ND_PRINT("\n\t\t%s subTLV (%u), length: %u, value: ",
+		 tok2str(lsa_opaque_ep_extd_prefix_subtlv_values,"unknown",subtlv_type),
+		 subtlv_type,
+		 subtlv_length);
+
+	switch (subtlv_type) {
+	case LS_OPAQUE_EP_SUBTLV_PREFIX_SID:
+	    flags = GET_U_1(tptr);
+	    mt_id = GET_U_1(tptr+2);
+	    algo = GET_U_1(tptr+3);
+
+	    if (subtlv_length == 7) {
+		ND_PRINT("\n\t\t  Label: %u, MT-ID: %u, Algorithm: %u",
+			 GET_BE_U_3(tptr+4), mt_id, algo);
+	    } else if (subtlv_length == 8) {
+		ND_PRINT("\n\t\t  Index: %u, MT-ID: %u, Algorithm: %u, Flags [%s]",
+			 GET_BE_U_4(tptr+4), mt_id, algo,
+			 bittok2str(ep_range_tlv_prefix_sid_subtlv_flag_values, "none", flags));
+	    } else {
+		ND_PRINT("\n\t\tBogus subTLV length %u", subtlv_length);
+	    }
+	    break;
+
+	default:
+	    if (ndo->ndo_vflag <= 1) {
+		if (!print_unknown_data(ndo, tptr, "\n\t\t", subtlv_length))
+		    return -1;
+	    }
+	}
+
+	/* in OSPF everything has to be 32-bit aligned, including subTLVs */
+	if (subtlv_length % 4) {
+	    subtlv_length += (4 - (subtlv_length % 4));
+	}
+	tptr+=subtlv_length;
+	tlv_length-=subtlv_length;
+    }
+    return 0;
+trunc:
+    return -1;
+}
+
+static int
+ospf_ep_lsa_print(netdissect_options *ndo, const uint8_t *tptr, u_int lsa_length)
+{
+    u_int tlv_type, tlv_length;
+    uint16_t range_size;
+    uint8_t af, prefix_length, route_type, flags;
+
+    while (lsa_length >= 4) {
+
+	ND_TCHECK_4(tptr);
+	tlv_type = GET_BE_U_2(tptr);
+	tlv_length = GET_BE_U_2(tptr+2);
+	tptr+=4;
+	lsa_length-=4;
+
+	/* Infinite loop protection. */
+	if (tlv_type == 0 || tlv_length == 0) {
```

Comment:
nd_print_invalid(ndo); before the return?

Reply:
do you mean I should do this also on the trunc: label ?

Reply:
ok

Reply:
`nd_print_invalid(ndo);` still missing.